src/async-syntax-highlighter.js

```diff
@@ -15,41 +17,84 @@ export default (options) => {
       return ReactAsyncHighlighter.loadAstGenerator();
     }
 
-    static registerLanguage = (_, language) => {
-      if (!supportsLanguageRegistering) {
+    static async loadLanguage(language) {
+      const languageLoader = languageLoaders[language];
+      if (typeof languageLoader === 'function') {
+        return languageLoader(ReactAsyncHighlighter.registerLanguage);
+      } else {
+        throw `Language ${language} not supported`
+      }
+    }
+    
+    static isRegistered = (language) => {
+      if(!registerLanguage) {
+        return true;
+      }
+
+      if (!ReactAsyncHighlighter.astGenerator) {
+        // Ast generator not available yet, but language will be registered once it is.
+        return ReactAsyncHighlighter.languages.findIndex(item => item.name === language) > -1;
+      }
+
+      return isLanguageRegistered(ReactAsyncHighlighter.astGenerator, language);
+    }
+
+    static registerLanguage = (name, language) => {
+      if (!registerLanguage) {
         return;
       }
       
       if(ReactAsyncHighlighter.astGenerator) {
-        return ReactAsyncHighlighter.astGenerator.register(language);
+        return registerLanguage(ReactAsyncHighlighter.astGenerator, name, language);
       } else {
-        ReactAsyncHighlighter.languages.push(language);
+        ReactAsyncHighlighter.languages.push({
+          name,
+          language
+        });
       }
     };
-    
 
     static loadAstGenerator() {
       ReactAsyncHighlighter.astGeneratorPromise = loader().then(astGenerator => {
         ReactAsyncHighlighter.astGenerator = astGenerator;
 
-        if (supportsLanguageRegistering && ReactAsyncHighlighter.languages.length) {
-          ReactAsyncHighlighter.languages.forEach((language) => astGenerator.register(language));
+        if (registerLanguage && ReactAsyncHighlighter.languages.length) {
+          ReactAsyncHighlighter.languages.forEach(({ name, language }) => registerLanguage(astGenerator, name, language));
         }
       });
       
       return ReactAsyncHighlighter.astGeneratorPromise;
     }
-  
+    
+    componentDidUpdate() {
+      if(!ReactAsyncHighlighter.isRegistered(this.props.language) && languageLoaders) {
+        this.loadLanguage();
+      }
+    }
+
     componentDidMount() {
       if (!ReactAsyncHighlighter.astGeneratorPromise) {
         ReactAsyncHighlighter.loadAstGenerator();
       }
-  
+      
       if(!ReactAsyncHighlighter.astGenerator) {
         ReactAsyncHighlighter.astGeneratorPromise.then(() => {
           this.forceUpdate();
         });
       }
+      if(!ReactAsyncHighlighter.isRegistered(this.props.language) && languageLoaders) {
+        this.loadLanguage();
+      }
+    }
+
+    async loadLanguage() {
+      try {
+        ReactAsyncHighlighter.loadLanguage(this.props.language).then(() => {
+          this.forceUpdate();
+        });
+      } catch (error) {
+        console.log(error);
```

Comment:
is there a better way to handle than console.log here?

Reply:
I'll remove the try catch, it should just throw all the way up